File: pkg/engine/plan.go
package engine

import (
	"fmt"
	"sort"
	"strings"

	"github.com/brandonapol/normal/pkg/config"
)

type ActionKind string

const (
	ActionWriteFile      ActionKind = "write-file"
	ActionDeleteFile     ActionKind = "delete-file"
	ActionRestartService ActionKind = "restart-service"
)

type Action struct {
	Kind     ActionKind `json:"kind"`
	Path     string     `json:"path,omitempty"`
	Contents string     `json:"contents,omitempty"`
	Service  string     `json:"service,omitempty"`
}

type HealthCheck struct {
	Service string `json:"service"`
}

type Plan struct {
	FromRevision int           `json:"fromRevision"`
	ToRevision   int           `json:"toRevision"`
	Intent       string        `json:"intent,omitempty"`
	ApprovedBy   string        `json:"approvedBy,omitempty"`
	DigestBefore string        `json:"digestBefore,omitempty"`
	DigestAfter  string        `json:"digestAfter,omitempty"`
	Diff         Diff          `json:"diff"`
	Actions      []Action      `json:"actions"`
	Services     []string      `json:"services"`
	Checks       []HealthCheck `json:"checks"`
}

func (p Plan) IsNoOp() bool { return len(p.Actions) == 0 }

type PlanIssue struct {
	Path    string `json:"path"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

type PlanError struct {
	Issues []PlanIssue
}

func (e *PlanError) Error() string {
	parts := make([]string, 0, len(e.Issues))
	for _, issue := range e.Issues {
		parts = append(parts, fmt.Sprintf("%s: %s", issue.Path, issue.Message))
	}
	return "plan rejected: " + strings.Join(parts, "; ")
}

func (e *PlanError) Codes() []string {
	codes := make([]string, 0, len(e.Issues))
	for _, issue := range e.Issues {
		codes = append(codes, issue.Code)
	}
	return codes
}

var immutablePaths = []string{"/apiVersion", "/kind"}

func diffFileSets(before, after FileSet) []Action {
	paths := make(map[string]bool, len(before)+len(after))
	for path := range before {
		paths[path] = true
	}
	for path := range after {
		paths[path] = true
	}
	ordered := make([]string, 0, len(paths))
	for path := range paths {
		ordered = append(ordered, path)
	}
	sort.Strings(ordered)

	writes := make([]Action, 0)
	deletes := make([]Action, 0)
	for _, path := range ordered {
		next, present := after[path]
		if !present {
			deletes = append(deletes, Action{Kind: ActionDeleteFile, Path: path})
			continue
		}
		if before[path] != next {
			writes = append(writes, Action{Kind: ActionWriteFile, Path: path, Contents: next})
		}
	}
	return append(writes, deletes...)
}

func planIssues(changes []Change) []PlanIssue {
	issues := make([]PlanIssue, 0)
	for _, change := range changes {
		immutable := false
		for _, path := range immutablePaths {
			if covers(path, change.Path) {
				immutable = true
				break
			}
		}
		if immutable {
			issues = append(issues, PlanIssue{
				Path:    change.Path,
				Code:    "immutable-field",
				Message: "apiVersion and kind cannot be changed by a mutation; migrate the config instead",
			})
			continue
		}
		if !IsOwnedPath(change.Path) {
			issues = append(issues, PlanIssue{
				Path:    change.Path,
				Code:    "unowned-path",
				Message: "no service owns this path, so the engine cannot apply it",
			})
		}
	}
	return issues
}

func PlanApply(current, desired config.Config) (Plan, error) {
	diff, err := DiffConfigs(current, desired)
	if err != nil {
		return Plan{}, err
	}

	if diff.IsEmpty() {
		return Plan{
			FromRevision: current.Metadata.Revision,
			ToRevision:   desired.Metadata.Revision,
			Diff:         diff,
			Actions:      []Action{},
			Services:     []string{},
			Checks:       []HealthCheck{},
		}, nil
	}

	issues := planIssues(diff.Changes)
	if desired.Metadata.Revision <= current.Metadata.Revision {
		issues = append(issues, PlanIssue{
			Path:    "/metadata/revision",
			Code:    "stale-revision",
			Message: fmt.Sprintf("revision must advance past %d", current.Metadata.Revision),
		})
	}
	if len(issues) > 0 {
		return Plan{}, &PlanError{Issues: issues}
	}

	currentFiles, err := Render(current)
	if err != nil {
		return Plan{}, err
	}
	desiredFiles, err := Render(desired)
	if err != nil {
		return Plan{}, err
	}

	fileActions := diffFileSets(currentFiles, desiredFiles)
	digestBefore := Digest(currentFiles)
	digestAfter := Digest(desiredFiles)
	touched := make([]string, 0, len(fileActions))
	for _, action := range fileActions {
		touched = append(touched, action.Path)
	}
	services := ServicesForFiles(touched)

	actions := make([]Action, 0, len(fileActions)+len(services))
	actions = append(actions, fileActions...)
	checks := make([]HealthCheck, 0, len(services))
	for _, service := range services {
		actions = append(actions, Action{Kind: ActionRestartService, Service: service})
		checks = append(checks, HealthCheck{Service: service})
	}

	return Plan{
		FromRevision: current.Metadata.Revision,
		ToRevision:   desired.Metadata.Revision,
		DigestBefore: digestBefore,
		DigestAfter:  digestAfter,
		Diff:         diff,
		Actions:      actions,
		Services:     services,
		Checks:       checks,
	}, nil
}

func WithNextRevision(current, desired config.Config) config.Config {
	next := desired
	next.Metadata.Revision = current.Metadata.Revision + 1
	return next
}

File: packages/engine/src/plan.ts
import { err, ok, type NormalConfig, type Result } from "@normal/schema";
import { diffConfig, type Change, type ConfigDiff } from "./diff.js";
import { isOwnedPath, servicesForFiles, type ServiceName } from "./ownership.js";
import { renderConfig, type FileSet } from "./render.js";

export type FileAction =
  | { readonly kind: "write-file"; readonly path: string; readonly contents: string }
  | { readonly kind: "delete-file"; readonly path: string };

export type Action = FileAction | { readonly kind: "restart-service"; readonly service: ServiceName };

export type HealthCheck = {
  readonly service: ServiceName;
};

export type Plan = {
  readonly fromRevision: number;
  readonly toRevision: number;
  readonly diff: ConfigDiff;
  readonly actions: readonly Action[];
  readonly services: readonly ServiceName[];
  readonly checks: readonly HealthCheck[];
};

export type PlanIssue = {
  readonly path: string;
  readonly code: "immutable-field" | "unowned-path" | "stale-revision";
  readonly message: string;
};

const IMMUTABLE_PATHS = ["/apiVersion", "/kind"];

const diffFileSets = (before: FileSet, after: FileSet): FileAction[] => {
  const paths = [...new Set([...Object.keys(before), ...Object.keys(after)])].sort();
  const writes: FileAction[] = [];
  const deletes: FileAction[] = [];
  for (const path of paths) {
    const previous = before[path];
    const next = after[path];
    if (next === undefined) {
      deletes.push({ kind: "delete-file", path });
      continue;
    }
    if (previous !== next) writes.push({ kind: "write-file", path, contents: next });
  }
  return [...writes, ...deletes];
};

const issuesFor = (changes: readonly Change[]): PlanIssue[] =>
  changes.flatMap((change): PlanIssue[] => {
    if (IMMUTABLE_PATHS.some((path) => change.path === path || change.path.startsWith(`${path}/`))) {
      return [
        {
          path: change.path,
          code: "immutable-field",
          message: "apiVersion and kind cannot be changed by a mutation; migrate the config instead",
        },
      ];
    }
    if (!isOwnedPath(change.path)) {
      return [
        {
          path: change.path,
          code: "unowned-path",
          message: "no service owns this path, so the engine cannot apply it",
        },
      ];
    }
    return [];
  });

export const planApply = (
  current: NormalConfig,
  desired: NormalConfig,
): Result<Plan, PlanIssue[]> => {
  const diff = diffConfig(current, desired);

  if (diff.changes.length === 0) {
    return ok({
      fromRevision: current.metadata.revision,
      toRevision: desired.metadata.revision,
      diff,
      actions: [],
      services: [],
      checks: [],
    });
  }

  const issues = issuesFor(diff.changes);
  if (desired.metadata.revision <= current.metadata.revision) {
    issues.push({
      path: "/metadata/revision",
      code: "stale-revision",
      message: `revision must advance past ${current.metadata.revision}`,
    });
  }
  if (issues.length > 0) return err(issues);

  const fileActions = diffFileSets(renderConfig(current), renderConfig(desired));
  const services = servicesForFiles(fileActions.map((action) => action.path));
  const restarts: Action[] = services.map((service) => ({ kind: "restart-service", service }));

  return ok({
    fromRevision: current.metadata.revision,
    toRevision: desired.metadata.revision,
    diff,
    actions: [...fileActions, ...restarts],
    services,
    checks: services.map((service) => ({ service })),
  });
};

export const isNoOp = (plan: Plan): boolean => plan.actions.length === 0;

export const nextRevision = (current: NormalConfig, desired: NormalConfig): NormalConfig => ({
  ...desired,
  metadata: { ...desired.metadata, revision: current.metadata.revision + 1 },
});
